rfcs/2022-03-07-3829-better-support-for-large-aws-s3-batches.md

```diff
@@ -0,0 +1,398 @@
+   order.
+3. The final object is assembled with the `CompleteMultipartUpload`
+   action.
+
+There are two major issues that this API presents that force some of
+the implementation decisions below:
+
+1. Vector may need to continue a multipart upload process across sink
+   restarts (due to reloads, restarts, or crashes). Since the uploads
+   require the presence of a unique identifier for both the multipart
+   upload itself as well as each of the uploaded parts, Vector will
+   need to store these identifiers with the partition that initiated
+   the upload.
+2. The upload parts (except for the final part) have a _minimum_ size
+   limit of 5MB. Since the sink may need to be restarted before a
+   batch reaches this minimum size, it will also need to buffer the
+   batch to disk before uploading the part.
+
+### User Experience
+
+The enhanced sink will add support for a new time-based upload
+mode. This mode will create a new object in the destination bucket
+once every configured interval instead of creating an object when a
+batch is filled or times out. This new object will be uploaded
+incrementally, but will not be visible in the bucket until all the
+parts of it are complete.
+
+The following upload behaviors are possible through the configuration
+described below:
+
+1. The sink operates as before, where batches are buffered based only
+   on the existing batch settings and uploaded directly to individual
+   objects. This will initially be the default configuration to avoid
+   breaking existing setups.
+1. The sink buffers batches until it can generate a complete part, and
+   assembles the parts into an object only at the end of an
+   interval. None of the intermediate parts are visible until the
+   final object is assembled.
+1. The sink uploads batches into parts as required to ensure the data
+   is persisted to S3, and assembles the parts into an object at the
+   end of an interval _or_ when restarting the upload would
+   not be possible due to an undersized part.
+
+This change will introduce the following new configuration items for
+this sink, which will all default to operating the same as the
+existing sink.
+
+- `data_dir` (string): The path to the directory in which the state
+  data is stored. If this is not set, it is derived from the global
+  `data_dir` setting.
+- `batch.min_bytes` (integer): This specifies the target size a
+  locally-stored part must reach before it is uploaded. Due to the
+  limitations of the S3 API, this has an minimum value of 5MB, which
+  is also the default value when `upload_interval` is set.
+- `upload_interval` (duration or enum): How often should the sink
+  complete the upload process and initiate a new one. If this is
+  unset, the existing batch-based object creation mechanism is
+  used. In addition to common interval specifications (ie `"15
+  minutes"`, etc), the following values are permitted:
+  - `"minutely"`: Complete the upload at the end of the minute exactly.
```

Comment:
This doesn't seem correct from an English perspective. I wonder if it'd be clearer to change these to something that's more explicit about it indicating a particular position in each minute/day/hour.

Reply:
Somewhat obvious answer of cron style syntax - though that definitely covers more than we need

Reply:
Alternatively we stick with our current pattern of specifying intervals in seconds.